docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+        Enclave2_VK []byte,     // creator of this message
+        Enclave_VK []byte,      // receiver of this message
+        Sig_Enclave []byte,     // signature of the message creator
+}
+```
+
+ERCC keeps in instance of an attestation.Verifier to verify the attestation evidence. In case of EPID, the evidence is an `IASAttestationReport` object. This instance has the Intel attestation verification key.
+```go
+type EnclaveRegistryCC struct {
+	ra  attestation.Verifier
+}
+
+// Intel IAS verification key (same as used by TLCC Enclave)
+const IntelPubPEM = `...`
+
+type IASAttestationReport struct {
+	EnclavePk                   []byte `json:"EnclavePk"`
+	IASReportSignature          string `json:"IASReport-Signature"`
+	IASReportSigningCertificate string `json:"IASReport-Signing-Certificate"`
+	IASReportBody               []byte `json:"IASResponseBody"`
+}
+```
+
+
+# TLCC
+
+## Interface:
+The TLCC interface is implemented through a "normal" chaincode interface.
+
+```go
+// chaincode interface 
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// TODO note that we are multiplexing normal and range queries with this method
+// used by ecc via cc2cc
+getMetadata(key string, nonce string, isRangeQuery bool) (cmacBase64  string, error)
+
+validateMSP(mspId string, nonce string) (valid bool, cmacBase64 string, error)
+
+// used by the admin to join the channel
+joinChannel(channel_id string) error
+```
+
+```c++
+// interface exposed to TLCC_enclave via cgo
+// logging
+void ocall_print_string([in, string] const char *str);
+```
+
+## State:
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// TBD is this interface already sufficient to support multi-channel?
+// TODO check if a block contatains any channel information?
+
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+
+// verify state data
+public int ecall_get_state_metadata(
+        [in, string] const char *key, // TBD: should we add namespace separetely?
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+public int ecall_get_multi_state_metadata(
+        [in, string] const char *comp_key,
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+// verify msp identities
+public int ecall_validate_msp_id(
+        [in, string] const char *msp_id,
+        [in, size=32] uint8_t *nonce,
+        [out] bool isValid, 
+        [out] sgx_cmac_128bit_tag_t *cmac);
```

Comment:
Wouldn't this correspond to the `can_endorse` from the UML diagrams? So why not use the same name?  The `can_endorse` seems also to capture more closely what we validate?

Reply:
you added can_endorse. in that case, do we still need validate_msp_id? what exactly would it do?

Reply:
What we _do_ need though additionally is something which helps in implementing `get_creator`.  Maybe easiest is a function which passes a SignedRequest and returns a (validated) tuple <msp-id, x509-principal/cert>`?

Reply:
mh ... passing the the entire proposal to tlcc to get back the creator is a bit too much. I suggest that we call this method `ecall_validate_identity`, and offer this to the chaincode as primitive to validate if the creator is valid or not.

Reply:
we definitely want to integrate this check into the shim's `GetCreator`. What we pass to tlcc i guess depends a bit on how much we already have to parse and decompose signed proposal.  your current validate_identity seems though not sufficient: you certainly would not send an identity string, at the very least you would need to send the certificate (from which ecc would have to extract the PK to verify the signature) and the corresponding MSP id it should match?

Reply:
good catch, `validate_identity` should take as input a serialzed identity as defined in fabric [here](https://github.com/hyperledger/fabric-protos/blob/master/msp/identities.proto#L15). I updated `validate_identity`

Reply:
using serialized identity is probably easier. Given that both sides have to unserialize it and we also support only X509, one could also just send the X509 cert plus msp id string.  But the duplicate deserialization is a non-issue and using serialized identity would match the fabric formats (even though i have to admit i hate their nested serializations ...)

Apropos "hate": is there anyway to tell github _not_ to to hide conversations which contains unresolved discussions?